### PR TITLE
Kernel user space communication

### PR DESCRIPTION
**Pull request by Shiranuit**
The Goal of this PR is to enable communication between Kernel Space and User Space applications.
This will allow us to communicate with the Lycanite driver from an application later on.

### CODE REVIEW

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 1]
```diff
@@ -298,20 +415,335 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+    perms |= (((UINT64)Input[13]) & 0xFF) << 32L;
+    perms |= (((UINT64)Input[14]) & 0xFF) << 40L;
+    perms |= (((UINT64)Input[15]) & 0xFF) << 48L;
+    perms |= (((UINT64)Input[16]) & 0xFF) << 56L;
+
+    len |= (((UINT16)Input[17]) & 0xFF);
+    len |= (((UINT16)Input[18]) & 0xFF) << 8L;
+
+    char* Message = (char*)calloc(len, sizeof(char));
+    RtlCopyMemory(Message, (PCHAR)Input+19, len);
```
Comment (Vaati3) [inline KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:536]: What happen if calloc return NULL?

Comment (cyrilbegard) [reply KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c]: fix

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 2]
```diff
@@ -298,20 +415,335 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+    perms |= (((UINT64)Input[5]) & 0xFF) << 32L;
+    perms |= (((UINT64)Input[6]) & 0xFF) << 40L;
+    perms |= (((UINT64)Input[7]) & 0xFF) << 48L;
+    perms |= (((UINT64)Input[8]) & 0xFF) << 56L;
+
+    len |= (((UINT16)Input[9]) & 0xFF);
+    len |= (((UINT16)Input[10]) & 0xFF) << 8L;
+
+    char* Message = (char*)calloc(len, sizeof(char));
+    RtlCopyMemory(Message, (PCHAR)Input + 11, len);
```
[the remark this answers is not in the document]
Comment (cyrilbegard) [reply KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:574]: fix

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 3]
```diff
@@ -298,20 +415,335 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+    pid |= (((UINT64)Input[5]) & 0xFF) << 32L;
+    pid |= (((UINT64)Input[6]) & 0xFF) << 40L;
+    pid |= (((UINT64)Input[7]) & 0xFF) << 48L;
+    pid |= (((UINT64)Input[8]) & 0xFF) << 56L;
+
+    len |= (((UINT16)Input[9]) & 0xFF);
+    len |= (((UINT16)Input[10]) & 0xFF) << 8L;
+
+    char* Message = (char*)calloc(len, sizeof(char));
+    RtlCopyMemory(Message, (PCHAR)Input + 11, len);
```
[the remark this answers is not in the document]
Comment (cyrilbegard) [reply KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:617]: fix

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 4]
```diff
@@ -298,20 +415,335 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+        return INVALID_REQUEST_SIZE;
+    }
+
+    UINT16 len = 0;
+
+    len |= (((UINT16)Input[1]) & 0xFF);
+    len |= (((UINT16)Input[2]) & 0xFF) << 8L;
+
+    char* Message = (char*)calloc(len, sizeof(char));
+    RtlCopyMemory(Message, (PCHAR)Input + 3, len);
```
[the remark this answers is not in the document]
Comment (cyrilbegard) [reply KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:642]: fix

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 5]
```diff
@@ -298,20 +415,335 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+    _In_ BOOLEAN Create
+) {
+    UINT64 ParentId = (UINT64)PParentId;
+    UINT64 ProcessId = (UINT64)PProcessId;
+    if (LPID != 0) {
+        if (Create) {
+            if (LPID == ParentId) {
+                char* ID = (char*)calloc(24, sizeof(char));
+                UINT8 idlen = ulong2chr(ProcessId, ID);
+                ProcessPerm* newperm = (ProcessPerm*)calloc(1, sizeof(ProcessPerm));
```
[the remark this answers is not in the document]
Comment (cyrilbegard) [reply KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:721]: fix

**Review by Vaati3 — DISMISSED** [review]

**Review by theocerutti — DISMISSED** [review]

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 6]
```diff
@@ -36,31 +39,80 @@ const FLT_OPERATION_REGISTRATION Callbacks[] = {
...
 
+UINT8 ulong2chr(UINT64 value, char* str) {
+    UINT8 pos = 0;
+    while (value != 0) {
+        str[pos++] = (value % 10) + '0';
+        value /= 10;
+    }
+
+    for (UINT8 i = 0; i < pos / 2; i++) {
+        char tmp = str[i];
```
Comment (AntoineMaillard06) [inline KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:79]: Doesn't better to init `char tmp` before **for** loop ?

File: KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c [hunk 7]
```diff
@@ -298,20 +415,351 @@ FLT_PREOP_CALLBACK_STATUS AvPreSetInformation(_Inout_ PFLT_CALLBACK_DATA Data, _
...
+        case SET_AUTHORIZATION_GLOBAL:
+            status = comSetAuthorizationPid(Input, InputBufferSize, ProcessInfos);
+            break;
+        case GET_PROCESS_STATS:
+            status = comSetAuthorizationPid(Input, InputBufferSize, ProcessInfos);
+            break;
+        case DELETE_AUTHORIZATION_PID:
+            status = comSetAuthorizationPid(Input, InputBufferSize, ProcessInfos);
+            break;
+        case DELETE_AUTHORIZATION_GLOBAL:
```
Comment (AntoineMaillard06) [inline KMDFLycaniteFileFilter/KMDFLycaniteFileFilter.c:690]: Can reduce all this cases to one.

**Review by theocerutti — APPROVED** [review]

**Review by Vaati3 — APPROVED** [review]

**Review by AntoineMaillard06 — APPROVED** [review]

**Review by yoanVESSIERE — APPROVED** [review]

**Review by mic16 — APPROVED** [review]

**Pull request merged**